packages/core/src/odata/common/request-builder/batch/batch-response-transformer.ts

```diff
@@ -0,0 +1,220 @@
+    parsedBatchResponse: (ResponseData[] | ResponseData)[]
+  ): (ErrorResponse | ReadResponse | WriteResponses)[] {
+    return parsedBatchResponse.map(responseData => {
+      if (Array.isArray(responseData)) {
+        return this.transformChangeSet(responseData);
+      }
+      return isHttpSuccessCode(responseData.httpCode)
+        ? this.transformRetrieveResponse(responseData)
+        : this.transformErrorResponse(responseData);
+    });
+  }
+
+  private transformRetrieveResponse(responseData: ResponseData): ReadResponse {
+    return {
+      ...responseData,
+      type: this.getConstructor(responseData.body)!,
+      as: asReadResponse(
+        responseData.body,
+        this.responseDataAccessor,
+        this.deserializer
+      ),
+      isSuccess: () => true
+    };
+  }
+
+  private transformErrorResponse(responseData: ResponseData): ErrorResponse {
+    return { ...responseData, isSuccess: () => false };
+  }
+
+  private transformChangeSetSubResponse(
+    responseData: ResponseData
+  ): WriteResponse {
+    return {
+      ...responseData,
+      type: this.getConstructor(responseData.body),
+      as: asWriteResponse(
+        responseData.body,
+        this.responseDataAccessor,
+        this.deserializer
+      )
+    };
+  }
+
+  private transformChangeSet(changesetData: ResponseData[]): WriteResponses {
+    return {
+      responses: changesetData.map(subResponseData =>
+        this.transformChangeSetSubResponse(subResponseData)
+      ),
+      isSuccess: () => true
+    };
+  }
+
+  /**
+   * Retrieve the constructor for a specific single response body.
+   * @param responseBody The body of a single response as an object.
+   * @param entityToConstructorMap Mapping between entity names and their respective constructors.
+   * @param responseDataAccessor Response data access module.
+   * @returns The constructor if found in the mapping, undefined otherwise.
+   */
+  private getConstructor(
+    responseBody: Record<string, any>
+  ): Constructable<EntityBase> | undefined {
+    const entityJson = this.responseDataAccessor.isCollectionResult(
+      responseBody
+    )
+      ? this.responseDataAccessor.getCollectionResult(responseBody)[0]
+      : this.responseDataAccessor.getSingleResult(responseBody);
+
+    const entityUri = entityJson?.__metadata?.uri;
+    if (entityUri) {
+      return this.entityToConstructorMap[
+        parseEntityNameFromMetadataUri(entityUri)
+      ];
+    }
+
+    logger.warn('Could not parse constructor from response body.');
+  }
+}
+
+/**
+ * Transform the parsed batch response.
+ * @param parsedBatchResponse Two dimensional list of parsed batch sub responses.
+ * @param entityToConstructorMap A map that holds the entity type to constructor mapping.
+ * @param responseDataAccessor Response data access module.
+ * @param deserializer Entity deserializer.
+ * @returns An array of parsed sub responses of the batch response.
+ */
+export function transformBatchResponse(
```

Comment:
with exporting this method you exposed two methods: this one which is the intended one to use I would say and the one on the transformer instance. Perhaps it makes sense to hide the one in the class or add a static public method in the class so that we only expose funcitonality once.

Reply:
Currently nothing from the batch module is exported to the root. Maybe it makes sense to keep it like this. It is quite difficult to test any of this when it is hidden. The typical problem. Also applies to your question below. I never know how to solve this in a "good" way.